src/core/models.py

```diff
@@ -20,34 +22,64 @@ class StateBase:
     """Base state object."""
 
     def __init__(
-        self, relation: Relation | None, component: Unit | Application, substrate: SUBSTRATES
+        self,
+        relation: Relation | None,
+        data_interface: Data,
+        component: Unit | Application,
+        substrate: SUBSTRATES,
```

Comment:
**todo:** Although I get what this is going for, I don't think it's necessary as we only ever need to implement `RelationState`, IF  we are OK with optional types.

I suggest we remove `StateBase`, and implement `RelationState` as:

```python
class RelationState:
    """Relation state object."""

    def __init__(
        self,
        relation: Relation | None,
        data_interface: Data,
        component: Unit | Application | None,
        substrate: SUBSTRATES,
    ):
        self.relation = relation
        self.data_interface = data_interface
        self.component = component
        self.substrate = substrate
        self.relation_data = self.data_interface.as_dict(self.relation.id) if self.relation else {}

    @property
    def data(self) -> MutableMapping:
        """Data representing the state."""
        return self.relation_data

    def update(self, items: dict[str, str]) -> None:
        """Writes to relation_data."""
        delete_fields = [key for key in items if not items[key]]
        update_content = {k: items[k] for k in items if k not in delete_fields}

        self.relation_data.update(update_content)

        for field in delete_fields:
            del self.relation_data[field]
```

Reply:
Note, the `component` can also be `None`.

Reply:
Same as https://github.com/canonical/zookeeper-operator/pull/129#discussion_r1568894833

Reply:
Of course `update` can be useful in generic usage, when implementation may differ.

(Regarding `data`, I think we have the same implementation for all our cllasses.)

Reply:
These are the lint warning we get in case `component` can be `None`.

Please let me know what is your further preference (adding corresponding checks or keep component so it can't be `None`)

```
WARNING: there is a new pyright version available (v1.1.345 -> v1.1.359).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

/home/ubuntu/repos/zookeeper-operator/src/charm.py
  /home/ubuntu/repos/zookeeper-operator/src/charm.py:282:50 - error: "name" is not a known member of "None" (reportOptionalMemberAccess)
  /home/ubuntu/repos/zookeeper-operator/src/charm.py:282:91 - error: "name" is not a known member of "None" (reportOptionalMemberAccess)
  /home/ubuntu/repos/zookeeper-operator/src/charm.py:401:69 - error: "name" is not a known member of "None" (reportOptionalMemberAccess)
/home/ubuntu/repos/zookeeper-operator/src/core/cluster.py
  /home/ubuntu/repos/zookeeper-operator/src/core/cluster.py:50:16 - error: Expression of type "Relation | None" cannot be assigned to return type "Relation"
    Type "Relation | None" cannot be assigned to type "Relation"
      "None" is incompatible with "Relation" (reportGeneralTypeIssues)
/home/ubuntu/repos/zookeeper-operator/src/core/models.py
  /home/ubuntu/repos/zookeeper-operator/src/core/models.py:258:35 - error: "name" is not a known member of "None" (reportOptionalMemberAccess)
  /home/ubuntu/repos/zookeeper-operator/src/core/models.py:310:38 - error: "name" is not a known member of "None" (reportOptionalMemberAccess)
  /home/ubuntu/repos/zookeeper-operator/src/core/models.py:310:89 - error: "name" is not a known member of "None" (reportOptionalMemberAccess)
```

Reply:
I think we can just pass `None` as the component, as we're not using it.

So in `RelationState` we have:
```python
        component: Unit | Application | None,
```

Reply:
And then, just pass `None` to the objects that don't need the component.

Reply:
What is your proposition to this case? I'm asking it to avoid returning any type that may not be preferred.

```
252     @property
253     def unit_id(self) -> int:
254         """The id of the unit from the unit name.
255 
256         e.g zookeeper/2 --> 2
257         """                                                                                                                                                                                                     
258         return int(self.component.name.split("/")[1])
```

Reply:
Do you want the same on

```
309         if self.substrate == "k8s":                  
310             host = f"{self.component.name.split('/')[0]}-{self.unit_id}.{self.component.name.split('/')[0]}-endpoints"   
```
?

Reply:
For the record: offline responded yes.

Reply:
Done. Code from above is copy-pasted as requested.